### PR TITLE
Task-37622: Implement Dlp feature V0

### PR DESCRIPTION
**Pull request by azayati**

### CODE REVIEW

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/processor/impl/DlpOperationProcessorImpl.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        DlpOperation dlpOperation = createDlpOperationsIterator.next();
+        connector.processItem(dlpOperation.getEntityId());
+        // Delete this single operation since it's not proceeded in bulk
+        dlpQueueSorted.remove(dlpOperation.getEntityId());
+      }
+    }
+
+    // Removes the processed IDs from the “dlp queue” table that have timestamp
+    // older than the timestamp of start of processing
+    dlpOperationDAO.deleteAllDlpOperationsHavingIdLessThanOrEqual(maxDlpOperationId);
```
Comment (rdenarie) [inline commons-dlp/src/main/java/org/exoplatform/commons/dlp/processor/impl/DlpOperationProcessorImpl.java]: I'm not sure to understand what we delete here. Can you explain ?

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/processor/impl/DlpOperationProcessorImpl.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+      while (createDlpOperationsIterator.hasNext()) {
+        if (isInterrupted()) {
+          throw new RuntimeException("Dlp queue processing interrupted");
+        }
+        DlpOperation dlpOperation = createDlpOperationsIterator.next();
+        connector.processItem(dlpOperation.getEntityId());
+        // Delete this single operation since it's not proceeded in bulk
+        dlpQueueSorted.remove(dlpOperation.getEntityId());
+      }
+    }
```
Comment (rdenarie) [inline commons-dlp/src/main/java/org/exoplatform/commons/dlp/processor/impl/DlpOperationProcessorImpl.java]: Can you explain what this for does ?

**Review by rdenarie — CHANGES_REQUESTED** [review]
We should add unit tests

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/dao/DlpOperationDAO.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import java.util.List;
+
+import org.exoplatform.commons.api.persistence.GenericDAO;
+import org.exoplatform.commons.dlp.domain.DlpOperation;
+
+public interface DlpOperationDAO extends GenericDAO<DlpOperation, Long> {
+
+  List<DlpOperation> findAllFirst(Integer maxResults);
+
+  void deleteAllDlpOperationsHavingIdLessThanOrEqual(long id);
```
Comment (rdenarie) [inline commons-dlp/src/main/java/org/exoplatform/commons/dlp/dao/DlpOperationDAO.java]: To remove I think

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/dao/impl/DlpOperationDAOImpl.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  @Override
+  @ExoTransactional
+  public void deleteAllDlpOperationsHavingIdLessThanOrEqual(long id) {
+    getEntityManager()
+        .createNamedQuery("DlpOperation.deleteAllDlpOperationsHavingIdLessThanOrEqual")
+        .setParameter("id", id)
+        .executeUpdate();
+  }
```
Comment (rdenarie) [inline commons-dlp/src/main/java/org/exoplatform/commons/dlp/dao/impl/DlpOperationDAOImpl.java]: to remove

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/domain/DlpOperation.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import javax.persistence.TemporalType;
+
+import org.exoplatform.commons.api.persistence.ExoEntity;
+
+@Entity
+@ExoEntity
+@Table(name = "DLP_QUEUE")
+@NamedQueries({
+    @NamedQuery(name = "DlpOperation.deleteAllDlpOperationsHavingIdLessThanOrEqual",
+        query = "DELETE FROM DlpOperation q WHERE q.id <= :id"),
```
Comment (rdenarie) [inline commons-dlp/src/main/java/org/exoplatform/commons/dlp/domain/DlpOperation.java]: To remove

File: commons-dlp/src/main/resources/conf/portal/configuration.xml [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+          <property name="groupName" value="Dlp"/>
+          <property name="job" value="org.exoplatform.commons.dlp.job.DlpJob"/>
+          <property name="repeatCount" value="0"/>
+          <property name="period" value="${exo.dlp.job.period:5000}"/><!-- default 5 seconds -->
+          <property name="startTime" value="+6000"/><!-- start after 1 minute delay -->
+          <property name="endTime" value=""/>
+        </properties-param>
+      </init-params>
+    </component-plugin>
+  </external-component-plugins>
```
Comment (rdenarie) [inline commons-dlp/src/main/resources/conf/portal/configuration.xml]: CronJob or PeriodJob ?

With a cronJob, we can say to Quartz to not start a new execution if the first one is not finished

In addition, I thinkg that execution time each 1 minute is sufficient

**Review by rdenarie — CHANGES_REQUESTED** [review]

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/domain/DlpOperation.java (hunk 5 above)
Comment (azayati) [reply commons-dlp/src/main/java/org/exoplatform/commons/dlp/domain/DlpOperation.java]: removed

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/dao/impl/DlpOperationDAOImpl.java (hunk 4 above)
Comment (azayati) [reply commons-dlp/src/main/java/org/exoplatform/commons/dlp/dao/impl/DlpOperationDAOImpl.java]: removed

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/dao/DlpOperationDAO.java (hunk 3 above)
Comment (azayati) [reply commons-dlp/src/main/java/org/exoplatform/commons/dlp/dao/DlpOperationDAO.java]: removed

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/connector/DlpServiceConnector.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Is extended by all DlpService connectors, and allows to build configuration needed by a list of connectors that is used for Dlp.
+ * 
+ */
+public abstract class DlpServiceConnector extends BaseComponentPlugin {
+  private String type; //type name
+  private String displayName; //for use when rendering
+  private boolean enable = true;
+  protected static final String DLP_FEATURE = "dlp";
+  protected static final String DLP_POSITIVE_DETECTION = "dlpPositiveDetection ";
```
Comment (rdenarie) [inline commons-dlp/src/main/java/org/exoplatform/commons/dlp/connector/DlpServiceConnector.java]:
```suggestion
  protected static final String DLP_POSITIVE_DETECTION = "dlpPositiveDetection";
```

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/connector/DlpServiceConnector.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Is extended by all DlpService connectors, and allows to build configuration needed by a list of connectors that is used for Dlp.
+ * 
+ */
+public abstract class DlpServiceConnector extends BaseComponentPlugin {
+  private String type; //type name
+  private String displayName; //for use when rendering
+  private boolean enable = true;
+  protected static final String DLP_FEATURE = "dlp";
```
Comment (rdenarie) [inline commons-dlp/src/main/java/org/exoplatform/commons/dlp/connector/DlpServiceConnector.java]: to be removed

**Review by rdenarie — CHANGES_REQUESTED** [review]

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/connector/DlpServiceConnector.java (hunk 7 above)
Comment (azayati) [reply commons-dlp/src/main/java/org/exoplatform/commons/dlp/connector/DlpServiceConnector.java]: done

File: commons-dlp/src/main/java/org/exoplatform/commons/dlp/connector/DlpServiceConnector.java (hunk 8 above)
Comment (azayati) [reply commons-dlp/src/main/java/org/exoplatform/commons/dlp/connector/DlpServiceConnector.java]: done

**Review by rdenarie — APPROVED** [review]

**Pull request merged**